### ISSUE TITLE
Error in WhoAmI On import and export start

### ISSUE THREAD

**Issue by bzuidgeest** [sorengranfeldt/psma#18]
Hello,

After some time in which the agent worked fine, we recently ran into an issue. According to the event log the agent was crashing in the whoAmI function of the agent. 

We ran debug view and this part was the relevant output:

[3880] PSMA Information: 0 : enter show-identity
[3880] PSMA Information: 0 : identity-name: digiworkz\*******
[3880] PSMA Information: 0 : identity-token: 4024
[3880] PSMA Information: 0 : identity-user-value: S-1-5-21-414************
[3880] PSMA Information: 0 : group-membership digiworkz\Domain Users
[3880] PSMA Information: 0 : group-membership Everyone
[3880] PSMA Information: 0 : group-membership BUILTIN\Users
[3880] PSMA Information: 0 : group-membership BUILTIN\Administrators
[3880] PSMA Information: 0 : group-membership NT AUTHORITY\SERVICE
[3880] PSMA Information: 0 : group-membership CONSOLE LOGON
[3880] PSMA Information: 0 : group-membership NT AUTHORITY\Authenticated Users
[3880] PSMA Information: 0 : group-membership NT AUTHORITY\This Organization
[3880] PSMA Information: 0 : group-membership LOCAL
[3880] PSMA Information: 0 : group-membership digiworkz\GL-Ser******
......
[3880] PSMA Information: 0 : group-membership digiworkz\DL-C***********
[3880] PSMA Error: -2146233087 : error-showing-current-identity, Some or all identity references could not be translated.
[3880] PSMA Information: 0 : exit show-identity
[3880] PSMA Error: -2146233087 : setupimpersonationtoken, Some or all identity references could not be translated.
[3880] PSMA Information: 0 : exit setupimpersonationtoken
[3880] PSMA Error: -2146233087 : getimportentries, Some or all identity references could not be translated.
[3880] PSMA Information: 0 : exit getimportentries

For a proper diagnosis I then recompiled the code and according to the "WhatEaxactlyAreYouDoing" property I found I will now explain what I was doing... :)

I lifted the code from whoami function into a console program to determine the offending group membership. The account has very heavy group nesting, so this was a lot easier. Turns out there is one group membership in there (a global group) that is not from the same domain as the service account. I will check with the systemoverlords to get it removed, but I am not sure that is possible.

I tried running this code on the desktop as a common user (using the service account identity) and everything resolves/tranlates fine. It only has a problem when running as part of the fim services it seems. I do not know why.

As this group information is only listed to the tracer and does not seem to be actually used, I wondered if you could add a try catch around the translate step so it will not crash the agent.

For example (from whoAmI function):

 ```
using (WindowsIdentity current = new WindowsIdentity("********@digiworkz.nl"))
                {
                    Tracer.TraceInformation("identity-name: {0}", (object)current.Name);
                    Tracer.TraceInformation("identity-token: {0}", (object)current.Token);
                    Tracer.TraceInformation("identity-user-value: {0}", (object)current.User.Value);
                    if (current.Actor != null)
                    {
                        Tracer.TraceInformation("identity-actor: {0}", (object)current.Actor.Name);
                        Tracer.TraceInformation("identity-actor-auth-type: {0}", (object)current.Actor.AuthenticationType);
                    }
                    if (current.Groups == null)
                        return;
                    foreach (IdentityReference group in current.Groups)
                    {
                       
                        try
                        {
                            Tracer.TraceInformation("group-membership {0}", (object)(group.Translate(typeof(NTAccount)) as NTAccount).Value);
                        }
                        catch
                        {
                            Tracer.TraceError("Error translating group {0}", group.Value);
// don't throw the error here.
                        }
                    }
                }
```

As the agent is open source I could do this myself, but I prefer to keep following the prepackaged version you provide and not deviate from "official".

p.s. sorry about the de-compilation, but it is a valuable diagnostic tool for me in cases like this. It also enables me to be specific and provide an actual ready solution instead of just asking people like you to find and fix it. I hope you understand that.

**Comment by sorengranfeldt** [comment]
This could certainly be done. I would accept a pull request for this. However, if I was you, I'd still wonder why this cannot be resolved in a certain context, that seems strange.

You can go ahead a create a pull request for the try/catch and I can include that for next release

**Comment by bzuidgeest** [comment]
I believe this resolving issue is normal if the nested group is from a different domain and a domain local group. Or something like that. The Token contains "all groups" but not all groups are in the global catalog. Then there are situations where the "trust" between domains might be one-way etc...

I should be able to make a pull request for you. That's fair enough.

**Comment by sorengranfeldt** [comment]
Super. I have a few things I'd like to bundle to new version and your try/catch could be included.

**Comment by sorengranfeldt** [comment]
Included in release 5.6.3.2021
